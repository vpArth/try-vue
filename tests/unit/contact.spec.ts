import ContactCard from '@/components/contact/Card.vue';
import ContactList from '@/components/contact/List.vue';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import Vuetify, { Vuetify as VuetifyInterface } from 'vuetify'
import './../jest-setup';

const localVue = createLocalVue();

describe('contact/', () => {
  let vuetify: VuetifyInterface;
  beforeEach(() => {
    vuetify = new Vuetify();
  });
  describe('Card.vue', () => {
    it('renders card with passed props', () => {
      const title = 'Contact Name';
      const phone = '+79998887766';
      const email = 'inbox@example.org';
      const descr = 'My contact';
      const wrapper = mount(ContactCard, {
        localVue,
        vuetify,
        propsData: { title, phone, email },
        slots:     { 'default': descr },
      });

      const cardTitle = wrapper.find('.v-card__title');
      expect(cardTitle.find('.card-title').text()).toBe(title);
      expect(cardTitle.find('.btn-phone .caption').text()).toBe('+7 (999) 888-77-66');
      expect(cardTitle.find('.btn-email .caption').text()).toBe(email);
      expect(cardTitle.find('.btn-phone').attributes('href')).toBe(`tel:${phone}`);
      expect(cardTitle.find('.btn-email').attributes('href')).toBe(`mailto:${email}`);
      expect(wrapper.find('.v-card__text').text()).toBe(descr);

      expect(wrapper.html()).toMatchSnapshot();

      // test remove button click
      const event = jest.fn();
      const rmBtn = wrapper.find('.btn-remove');
      wrapper.vm.$on('remove-btn:clicked', event);
      expect(event).toHaveBeenCalledTimes(0);
      rmBtn.trigger('click');
      expect(event).toHaveBeenCalledTimes(1);
    });
  });
  describe('List.vue', () => {
    it('renders empty list', () => {
      const wrapper = mount(ContactList, {
        localVue,
        vuetify,
        propsData: {
          cards: [],
        },
      });
      expect(wrapper.find('[role=list]').element.innerHTML.trim()).toStrictEqual('');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('List.vue', () => {
    let wrapper: Wrapper<ContactList>;
    beforeEach(() => {
      wrapper = mount(ContactList, {
        localVue,
        vuetify,
        propsData: {
          cards: [
            { title: 'Contact 1', phone: '+79998887766', email: 'one@example.org', description: 'Contact 1' },
            { title: 'Contact 2', phone: '+79996667788', email: 'two@example.org', description: 'Contact 2' },
          ],
        },
      });
    });
    it('renders two cards', () => {
      expect(wrapper.findAll('[role=listitem]').length).toBe(2);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('removes a card when cardRemove called and confirmed', () => {
      window.confirm = jest.fn().mockImplementation(() => true);
      // @ts-ignore
      wrapper.vm.cardRemove(0);
      wrapper.vm.$forceUpdate();
      expect(wrapper.props().cards.length).toBe(1);
      expect(wrapper.findAll('[role=listitem]').length).toBe(1);
    });
    it('leaves a card when cardRemove called and not confirmed', () => {
      window.confirm = jest.fn().mockImplementation(() => false);
      // @ts-ignore
      wrapper.vm.cardRemove(0);
      wrapper.vm.$forceUpdate();
      expect(wrapper.props().cards.length).toBe(2);
      expect(wrapper.findAll('[role=listitem]').length).toBe(2);
    });
    it('handles Card.remove-btn:clicked event (mock)', () => {
      // @ts-ignore
      wrapper.vm.cardRemove = jest.fn();
      const card = wrapper.find('.v-card');
      card.vm.$emit('remove-btn:clicked');
      // @ts-ignore
      expect(wrapper.vm.cardRemove).toBeCalledWith(0);
    });
    it('handles Card.remove-btn:clicked event (real)', () => {
      window.confirm = jest.fn().mockImplementation(() => true);
      const card = wrapper.find('.v-card');
      card.vm.$emit('remove-btn:clicked');
      wrapper.vm.$forceUpdate();
      expect(wrapper.props().cards.length).toBe(1);
    });
  });
});
