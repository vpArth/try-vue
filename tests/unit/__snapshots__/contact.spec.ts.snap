// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact/ Card.vue renders card with passed props 1`] = `
<div class="v-card v-sheet theme--light">
  <div class="v-card__title">
    <h2 class="card-title">Contact Name</h2>
    <div class="spacer"></div> <a href="tel:+79998887766" class="mx-2 btn-phone v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">phone</i> <span class="caption">+7 (999) 888-77-66</span></span></a> <a href="mailto:inbox@example.org" class="mx-2 btn-email v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">email</i> <span class="caption">inbox@example.org</span></span></a> <button type="button" class="mx-2 btn-remove v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light error--text">delete</i></span></button>
  </div>
  <div class="v-card__text">My contact</div>
</div>
`;

exports[`contact/ List.vue renders empty list 1`] = `<div role="list" class="v-list v-sheet v-sheet--tile theme--light"></div>`;

exports[`contact/ List.vue renders two cards 1`] = `
<div role="list" class="v-list v-sheet v-sheet--tile theme--light">
  <div tabindex="-1" role="listitem" class="v-list-item theme--light">
    <div class="v-list-item__content">
      <div class="v-card v-sheet theme--light">
        <div class="v-card__title">
          <h2 class="card-title">Contact 1</h2>
          <div class="spacer"></div> <a href="tel:+79998887766" class="mx-2 btn-phone v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">phone</i> <span class="caption">+7 (999) 888-77-66</span></span></a> <a href="mailto:one@example.org" class="mx-2 btn-email v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">email</i> <span class="caption">one@example.org</span></span></a> <button type="button" class="mx-2 btn-remove v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light error--text">delete</i></span></button>
        </div>
        <div class="v-card__text">Contact 1
        </div>
      </div>
    </div>
  </div>
  <div tabindex="-1" role="listitem" class="v-list-item theme--light">
    <div class="v-list-item__content">
      <div class="v-card v-sheet theme--light">
        <div class="v-card__title">
          <h2 class="card-title">Contact 2</h2>
          <div class="spacer"></div> <a href="tel:+79996667788" class="mx-2 btn-phone v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">phone</i> <span class="caption">+7 (999) 666-77-88</span></span></a> <a href="mailto:two@example.org" class="mx-2 btn-email v-btn v-btn--depressed v-btn--flat v-btn--outlined v-btn--tile theme--light v-size--default success--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left material-icons theme--dark">email</i> <span class="caption">two@example.org</span></span></a> <button type="button" class="mx-2 btn-remove v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light error--text">delete</i></span></button>
        </div>
        <div class="v-card__text">Contact 2
        </div>
      </div>
    </div>
  </div>
</div>
`;
